### ISSUE TITLE
Filterable() on BelongsToMany fields

### ISSUE THREAD

**Issue by jimarick** [laravel/nova-issues#3900]
When using filterable() on a BelongsToMany field, it does not work as expected. A filter is applied, which has a list of the related resource as options. However when an option is chosen, only one result is returned, and that is the result with an ID number that matches the ID of the chosen filter value, rather than all results related to that filter value. 

For example, I have two models called Question and Set. Each question can be in many sets, and each set can have many questions.

The Question model has a belongsToMany method called set:
 ```
public function Set()
    {
        return $this->belongsToMany(Set::class);
    }
```

The set model has the equivalent method:
```
    public function Questions()
    {
        return $this->belongsToMany(Question::class);
    }
```

The Question Nova Resource has:
```
 public function fields(NovaRequest $request)
    {
        return [ 
                ID::make()->sortable(),

                BelongsToMany::make('Set')->filterable()
        ];
    }
```
A filter is applied as expected, and you get the correct options within the filter representing the IDs of the related sets. However, when you select a set only one item is returned, and that item is not actually in the set, it just has an ID matching the ID of the set you choose.

<img width="927" alt="Screenshot 2022-04-08 at 07 17 28" src="https://user-images.githubusercontent.com/67220144/162376083-18385a0c-cb10-44ce-ae9c-4a6559a25ec6.png">

This is obviously not the behaviour that would be expected when applying a filter to a BelongsToMany Field. I would expect to choose a Set, and return a list of all the questions in that Set (there are 3 questions in this particular set). 

I have worked around this by creating a custom filter called Set, with the following Apply and Options methods:
```
 public function apply(NovaRequest $request, $query, $value)
    {
        return $query
            ->whereHas('set', function (Builder $query) use ($value) {
                $query->where('set_id', $value);
            })
             ->orderBy('id', 'asc');
    }
```
 ```
public function options(NovaRequest $request)
    {
        $sets = \App\Models\Set::all();
        return $sets->pluck('id');
    }
```
This filter now works as I would expect, showing all 3 questions within this set: 
<img width="928" alt="Screenshot 2022-04-08 at 07 24 00" src="https://user-images.githubusercontent.com/67220144/162376856-17ad3677-8b2b-4f20-b429-7d598554afce.png">

Finally, a BelongsToMany field is not represented on the index of the question resource. I have got around this by adding the following field to the question resource:
```
            Text::make('Set', function() {
                return $this->set->pluck('id')->implode(',') ?? null;
            })->onlyOnIndex(),
```
<img width="911" alt="Screenshot 2022-04-08 at 07 27 04" src="https://user-images.githubusercontent.com/67220144/162377207-af7531da-cfff-42f3-93d9-e537a16dd713.png">

I wonder if this could potentially be applied when you add showOnIndex() to a BelongsToMany field type? This showOnIndex() method currently seems to have no effect on a BelongsToMany Field. 

Hope this all makes sense,
Thanks

**Comment by jimarick** [comment]
@davidhemphill Saw this was closed and marked as completed, but not in the recent release (4.1.1). Will it be in the next release? Thanks

**Comment by crynobone** [comment]
Yes, next release.

**Comment by jimarick** [comment]
Hi, this particular issue does not appear to be fixed, despite the recent update to Nova 4.2.1.

I will explain in more detail below:

New laravel project with Nova installed

Create two models with belongsToMany relationships

```javascript
namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Question extends Model
{
    use HasFactory;

    public function Sets()
    {
        return $this->belongsToMany(Set::class)->withPivot('order');
    }
}
```


```javascript
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Set extends Model
{
    use HasFactory;

    public function Questions()
    {
        return $this->belongsToMany(Question::class)->withPivot('order');
    }
}

```


Create Corresponding Migrations including Pivot table

```javascript
<?php

use Illuminate\Database\Migrations\Migration;
use Illuminate\Database\Schema\Blueprint;
use Illuminate\Support\Facades\Schema;

return new class extends Migration
{
    public function up()
    {
        Schema::create('questions', function (Blueprint $table) {
            $table->id();
            $table->boolean('normal')->nullable();
            $table->text('finding')->nullable();
            $table->timestamps();
        });
    }
};

```


```javascript
<?php

use Illuminate\Database\Migrations\Migration;
use Illuminate\Database\Schema\Blueprint;
use Illuminate\Support\Facades\Schema;

return new class extends Migration
{
    public function up()
    {
        Schema::create('sets', function (Blueprint $table) {
            $table->id();
            $table->timestamps();
        });
    }
};

```


```javascript
<?php

use Illuminate\Database\Migrations\Migration;
use Illuminate\Database\Schema\Blueprint;
use Illuminate\Support\Facades\Schema;

return new class extends Migration
{
    public function up()
    {
        Schema::create('question_set', function (Blueprint $table) {
            $table->id();
            $table->unsignedBigInteger('question_id');
            $table->unsignedBigInteger('set_id');
            $table->integer('order')->nullable();
            $table->timestamps();
        });
    }
};

```


Populate Database

![image](https://res.craft.do/user/full/2dfa1c5b-0204-ac16-f54f-2a27cbaac6d6/doc/9E64D102-C0C9-41EE-849F-4F4F72A3D3D3/5C1D54CA-1C00-42D1-BFE3-5FE860CC0026_2/LFVSlxZ5cl1jGe1rdiBvnaR3eyj7GJY9a9X1x5Ev4AUz/Image.png)

![image](https://res.craft.do/user/full/2dfa1c5b-0204-ac16-f54f-2a27cbaac6d6/doc/9E64D102-C0C9-41EE-849F-4F4F72A3D3D3/A9F1AECF-8F5E-43D8-BD00-7FA1E3AC2A97_2/2gYNCc8S7OpW6gTnQPRX8OI4lEzzpl9KMXzu6EZ9sD0z/Image.png)

![image](https://res.craft.do/user/full/2dfa1c5b-0204-ac16-f54f-2a27cbaac6d6/doc/9E64D102-C0C9-41EE-849F-4F4F72A3D3D3/AD32993D-D8F0-484D-841B-DA6A821161B7_2/ZAjftujdvxp835Pty6oQ9ah22CMC6TGd2b7MfdkCMBsz/Image.png)

Create Corresponding Resources

```javascript
<?php

namespace App\Nova;

use Laravel\Nova\Fields\ID;
use Illuminate\Http\Request;
use Laravel\Nova\Fields\Text;
use Laravel\Nova\Fields\Boolean;
use Laravel\Nova\Fields\BelongsToMany;
use Laravel\Nova\Http\Requests\NovaRequest;

class Question extends Resource
{
    public static $model = \App\Models\Question::class;
    public static $title = 'id';
    public static $search = [
        'id',
    ];

    public function fields(NovaRequest $request)
    {
        return [
            BelongsToMany::make('Sets')

				  ->filterable()

                ->fields(function ($request, $relatedModel) {
                    return [
                        Text::make('Order'),
                    ];
                }),

            ID::make()->sortable(),

            Text::make('Finding'),

            Boolean::make('Normal'),
        ];
    }
}
```


```javascript
<?php

namespace App\Nova;

use Laravel\Nova\Fields\ID;
use Illuminate\Http\Request;
use Laravel\Nova\Fields\Text;
use Laravel\Nova\Fields\BelongsToMany;
use Laravel\Nova\Http\Requests\NovaRequest;

class Set extends Resource
{

    public static $model = \App\Models\Set::class;
    public static $title = 'id';
    public static $search = [
        'id',
    ];

    public function fields(NovaRequest $request)
    {
        return [
            ID::make()->sortable(),

            BelongsToMany::make('Questions')
                ->fields(function ($request, $relatedModel) {
                    return [
                        Text::make('Order'),
                    ];
                })
        ];
    }
}


```


Then open the app in the browser, and goto the Sets resource and look at set 5. It contains 5 Questions (as expected)

![image](https://res.craft.do/user/full/2dfa1c5b-0204-ac16-f54f-2a27cbaac6d6/doc/9E64D102-C0C9-41EE-849F-4F4F72A3D3D3/7D477873-5EA8-43D8-BB59-38F3105868D3_2/lHnqjyAupZ4b1M9hYl8WZxiXPy5q3iSNBq3Dk895BUEz/Image.png)

But, if you open up the Questions resource and filter by Set 5, only the question with the ID 5 appears. The other related questions are not visible.

![image](https://res.craft.do/user/full/2dfa1c5b-0204-ac16-f54f-2a27cbaac6d6/doc/9E64D102-C0C9-41EE-849F-4F4F72A3D3D3/E6DFAC85-303F-46B0-8E28-985384CBCABB_2/HLCD6N7FcyMWhqL5aRAxlxJ857ddY9JkB6T1EyyZ3V4z/Image.png)

Even if I take the question with ID 5 out of Set 5, then it still appears in the list, despite not even being related. It is just shown because it has the same ID as the Set which is filtered.

![image](https://res.craft.do/user/full/2dfa1c5b-0204-ac16-f54f-2a27cbaac6d6/doc/9E64D102-C0C9-41EE-849F-4F4F72A3D3D3/F14513B2-47EA-48C5-8213-C4BA012E6FBB_2/l7eq4mi4fYMa79JajYGwIukvqYM3b38PWlHy8imlb5oz/Screenshot%202022-04-15%20at%2008.42.54.png)

![image](https://res.craft.do/user/full/2dfa1c5b-0204-ac16-f54f-2a27cbaac6d6/doc/9E64D102-C0C9-41EE-849F-4F4F72A3D3D3/38BE09F4-5327-4F0E-9F13-1459F91B03F1_2/4eE8WMz96Rf2hPtcv38ZNJJ7xZA3yo4eNHsmHqmGP6sz/Screenshot%202022-04-15%20at%2008.43.57.png)

**Comment by crynobone** [comment]
Yes just notice this too. Currently working on a fix.
